mamedov_antyukhov/graph.cpp

```diff
@@ -2,101 +2,86 @@
 
-  if (first_vertex_id == second_vertex_id) {
-    color = Graph::Edge::Color::Green;
-    connections_[first_vertex_id].push_back(new_edge_id);
-    edges_.insert({new_edge_id, Edge(new_edge_id, first_vertex_id,
-                                     second_vertex_id, color)});
-
-    return new_edge_id;
+  if (second_vertex_depth == 1) {
+    update_vertex_depth(second_vertex_id, first_vertex_depth + 1);
   }
+  const auto color = get_edge_color(first_vertex_id, second_vertex_id);
+  const EdgeId new_edge_id = get_new_edge_id();
 
-  const auto first_vertex_depth = get_vertex_depth_in_basis(first_vertex_id);
-  const auto second_vertex_depth = get_vertex_depth_in_basis(second_vertex_id);
-
-  if (!second_vertex_depth) {
-    color = Graph::Edge::Color::Grey;
-    if (!first_vertex_depth) {
-      set_vertex_depth(first_vertex_id, 1);
-      set_vertex_depth(second_vertex_id, 2);
-    } else {
-      set_vertex_depth(second_vertex_id, first_vertex_depth + 1);
-    }
-  } else {
-    if (second_vertex_depth - first_vertex_depth == 1)
-      color = Graph::Edge::Color::Yellow;
-    else if (second_vertex_depth - first_vertex_depth == 2) {
-      color = Graph::Edge::Color::Red;
-      set_vertex_depth(second_vertex_id, first_vertex_depth + 1);
-    } else
-      std::runtime_error("Failed to determine color");
+  if (first_vertex_id != second_vertex_id) {
+    connections_[second_vertex_id].emplace_back(new_edge_id);
   }
-
-  connections_[first_vertex_id].push_back(new_edge_id);
-  connections_[second_vertex_id].emplace_back(new_edge_id);
+  connections_[first_vertex_id].emplace_back(new_edge_id);
 
   edges_.insert({new_edge_id,
                  Edge(new_edge_id, first_vertex_id, second_vertex_id, color)});
 
   return new_edge_id;
-};
-
-Graph::Depth Graph::get_vertex_depth_in_basis(VertexId vertex_id) const {
-  if (vertices_depth_.find(vertex_id) != vertices_depth_.end()) {
-    return vertices_depth_.at(vertex_id).first;
-  }
-  return 0;
 }
 
 Graph::Depth Graph::get_vertex_depth(VertexId vertex_id) const {
   assert(has_vertex_id(vertex_id));
   if (vertices_depth_.find(vertex_id) != vertices_depth_.end()) {
-    return vertices_depth_.at(vertex_id).second;
+    return vertices_depth_.at(vertex_id);
   }
   return 0;
 }
 
 void Graph::set_vertex_depth(VertexId vertex_id, Depth new_vertex_depth) {
-  const auto old_vertex_depth = get_vertex_depth(vertex_id);
-  if (get_vertex_depth_in_basis(vertex_id)) {
-    if (old_vertex_depth > new_vertex_depth) {
-      vertices_depth_.at(vertex_id).second = new_vertex_depth;
-      const auto& edge_ids = get_connected_edge_ids(vertex_id);
-      const auto& edges = get_edges();
-      for (auto edge : edge_ids) {
-        set_vertex_depth(edges.at(edge).get_second_vertex_id(),
-                         new_vertex_depth + 1);
-      }
-    }
+  vertices_depth_[vertex_id] = new_vertex_depth;
+  --new_vertex_depth;
+  if (new_vertex_depth < depth_levels_.size()) {
```

Comment:
Очень неочевидная логика
```suggestion
  if (new_vertex_depth < depth_levels_.size() - 1) {
```

Reply:
В этой функции потом опять используется переменная new_vertex_depth для записи в вектор, тогда и там придется на один уменьшать. Лучше, наверное, один раз убавить